### PR TITLE
Add No Props Destructure rule issue #143

### LINKED ISSUE #143: add rule (rrd) props destructed
```js
import { ref } from 'vue';

const { firstName, lastName } = defineProps({
  firstName: String,
  lastName: String,
});
```

Destructing props break reactivity.

### PR DESCRIPTION
**Pull request by sagi-daniel**
### Summary
I have created a new rule that checks the practice of props destructuring in Vue components. The goal of this rule is to ensure code readability, maintainability, and functionality by preventing issues arising from improper handling of props changes.

I also prepared documentation that details the problems associated with props destructuring and its impact on Vue components, following the style of other documentation examples.

### Description
Basic Destructuring Check: Verifies that props destructuring is not done directly with defineProps, which can lead to reactivity issues.

### Related Issues
Fixes #143 

### Type of Change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Documentation update

### CODE REVIEW

**Comment by David-Pena** [comment]
I think you have a a different linter setup? 🤔 @rrd108

**Comment by sagi-daniel** [comment]
I have no idea, I'll check my vs code tomorrow.

**Comment by David-Pena** [comment]
No no I deleted my comment. It was my mistake 🫡

**Comment by rrd108** [comment]
@sagi-daniel I have a few requests and comments about the PR.

Are you ok with English or should I do it in Hungarian?

**Comment by sagi-daniel** [comment]
> @sagi-daniel I have a few requests and comments about the PR.
> 
> Are you ok with English or should I do it in Hungarian?

Yes, totaly. We can speak in english. It s good opportunity improve my soft skills and hard skills at the same time.

**Comment by rrd108** [comment]
I didn't commented yet as was waiting for your answer.

The first thing:
When you open the project it supposed to ask you to install the recommended extensions. One of them is eslint.

It helps us to follow the same formatting.

Please install it. 

When you open any of the files you changed and do formatting it should remove the semicolumns.

Now it is harder to review the pr because I do not see what is code change at what is code change.

After you fixed it you should just commit the changes, it will update the pr and I will review that.

I am in a car this afternoon so I may need some time to check it.

And the most important part, thanks for the contribution!

**Comment by sagi-daniel** [comment]
> I didn't commented yet as was waiting for your answer.
> 
> The first thing: When you open the project it supposed to ask you to install the recommended extensions. One of them is eslint.
> 
> It helps us to follow the same formatting.
> 
> Please install it.
> 
> When you open any of the files you changed and do formatting it should remove the semicolumns.
> 
> Now it is harder to review the pr because I do not see what is code change at what is code change.
> 
> After you fixed it you should just commit the changes, it will update the pr and I will review that.
> 
> I am in a car this afternoon so I may need some time to check it.
> 
> And the most important part, thanks for the contribution!

Thank you! I m going to install it, I thought npm install added all necessary depedency... maybe i didnt recognize some npm warning while the project initialization.

question: Do I have to make a new branch before the PR? or just push it to the main?

**Comment by rrd108** [comment]
It is not npm, they are vscode extensions.

You can create a new branch, but not necesarry as the PR will be merged in one step anyway.

File: docs/.vitepress/config.ts [hunk 1]
```diff
@@ -3,18 +3,16 @@ import { defineConfig } from 'vitepress'
 // https://vitepress.dev/reference/site-config
 export default defineConfig({
   title: 'Vue Mess Detector',
-  description: 'A static code analysis tool for detecting code smells and best practice violations in **Vue** and **Nuxt** projects.',
+  description:
+    'A static code analysis tool for detecting code smells and best practice violations in **Vue** and **Nuxt** projects.',
```
Comment (rrd108) [inline docs/.vitepress/config.ts:7]: There is still some formatting issue. In this file you have added one line but practically all lines marked as changed.

Can you take a look?

File: docs/rules/rrd/no-props-destructure.md [hunk 2]
```diff
@@ -0,0 +1,65 @@
+# Props Destructuring
+
+Check if props are being destructured in a way that might affect readability, maintainability, or functionality.
```
Comment (rrd108) [inline docs/rules/rrd/no-props-destructure.md:3]: Please mention the problem here. 

Destructing break reactivity

File: docs/rules/rrd/no-props-destructure.md [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+Destructuring props can make the code cleaner by allowing direct access to individual properties:
+
+```
+const { someProp, anotherProp } = defineProps();
+
+```
+
+### Vue Specific Considerations:
+
+While destructuring can simplify code in many cases, it can lead to issues with Vue's reactivity system. When props are destructured, Vue may not be able to track changes to the destructured properties. This is because destructuring creates separate variables that do not maintain a reactive relationship with the original props object.
```
Comment (rrd108) [inline docs/rules/rrd/no-props-destructure.md:20]: Not "may" happen. It will happens

File: docs/rules/rrd/no-props-destructure.md [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+As a result:
+
+- Reactivity Issues: Changes to the original props may not be reflected in the destructured variables, potentially leading to outdated or incorrect data being used in the component.
+- Data Flow: Destructuring can obscure the data flow within the component, making it harder to trace how props are being used or modified.
+
+## ❓ Why it's good to follow this rule?
+
+- Readability: Destructuring props can make the code cleaner and more readable by directly accessing individual properties. This can help in understanding the component’s requirements and logic more quickly.
```
Comment (rrd108) [inline docs/rules/rrd/no-props-destructure.md:29]: Here you should state more clearly that destructing is not recommended at all.

File: docs/rules/rrd/no-props-destructure.md [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+- Consistency: Using destructuring consistently helps to align with the component’s requirements. However, inconsistent destructuring or using it inappropriately might lead to code that is hard to follow and debug.
+- Avoid Overuse: Destructuring too many props or doing it in complex ways might lead to code that’s harder to manage. It’s important to balance between simplifying code and keeping it understandable.
+  😱 Examples of code for which this rule will throw a warning
+
+::: warning
+The following code destructures the items prop directly inside the setup function without any additional handling. This can lead to confusion about where items are coming from and how they are used.
+:::
+
+```vue
+<script setup>
```
Comment (rrd108) [inline docs/rules/rrd/no-props-destructure.md:40]: Nice and clear example but defineProps missing its parameters

File: docs/rules/rrd/no-props-destructure.md [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+```
+
+## 🤩 How to fix it?
+
+::: tip Refactor the code to ensure that props are handled appropriately. You can either destructure props in a way that makes their usage clear or handle them in a more controlled manner. Ensure that destructuring is
+done consistently and in a way that enhances code readability and maintainability.
+:::
+
+```vue
+<script setup>
```
Comment (rrd108) [inline docs/rules/rrd/no-props-destructure.md:56]: Please use the same example as in the other code above

File: src/rules/rrd/noPropDestructure.ts [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+import type { FileCheckResult, Offense } from '../../types'
+
+const results: FileCheckResult[] = []
+
+const checkNoPropDestructure = (script: SFCScriptBlock | null, filePath: string) => {
+  if (!script) {
+    return
+  }
+
+  const regex = /(?:const|let|var)\s*\{\s*([^\}]+)\s*\}\s*=\s*defineProps\s*\(\s*\)/g
```
Comment (rrd108) [inline src/rules/rrd/noPropDestructure.ts:14]: You can remove var here. It will offend a new rule coming

File: src/rules/rules.ts [hunk 8]
```diff
@@ -35,15 +23,17 @@ export const RULES = {
...
     'plainScript',
     'propsDrilling',
     'scriptLength',
     'shortVariableName',
     'tooManyProps',
-    'vForWithIndexKey'
-  ]
-};
+    'vForWithIndexKey',
+    'noPropDestructure',
```
Comment (rrd108) [inline src/rules/rules.ts:34]: Keep the rules in alphabetic order

File: src/rulesReport.ts [hunk 9]
```diff
@@ -99,17 +100,18 @@ export const reportRules = (groupBy: GroupBy) => {
...
 
   // Output the report grouped by file
   Object.keys(offensesGrouped).forEach((key) => {
     console.log(`\n - ${key}`)
     offensesGrouped[key].forEach((offense) => {
       const isError = offense.message.includes(BG_ERR)
       // if health already has the file, push the error
-      if (health.some(h => h.file === offense.file)) {
-        const foundHealth = health.find(h => h.file === offense.file)
+      if (health.some((h) => h.file === offense.file)) {
```
Comment (rrd108) [inline src/rulesReport.ts:113]: At `(h)` the bracets can be omitted. Also some formatting issue

**Review by rrd108 — CHANGES_REQUESTED** [review]

**Comment by David-Pena** [comment]
I think its important to add that is possible to destructure `props` but using `toRefs()`

Example
```
<script setup>
const props = defineProps<{ name: string }>()
const { name } = toRefs(props) // It will keep reactivity
</script>
```

What do you think? @rrd108 @sagi-daniel

**Comment by rrd108** [comment]
> I think its important to add that is possible to destructure `props` but using `toRefs()`
> 
> Example
> 
> ```
> <script setup>
> const props = defineProps<{ name: string }>()
> const { name } = toRefs(props) // It will keep reactivity
> </script>
> ```
> 
> What do you think? @rrd108 @sagi-daniel

`toRefs` approach is worth a dedicated rule. Keep this one here clear and simple.

**Review by rrd108 — APPROVED** [review]
